dsl/DefaultArgs.cc

```diff
@@ -0,0 +1,168 @@
+    }
+
+    while (send != nullptr) {
+        switch (send->fun._id) {
+            case core::Names::params()._id: {
+                if (send->args.size() != 1) {
+                    return ast::MK::EmptyTree();
+                }
+                auto *hash = ast::cast_tree<ast::Hash>(send->args[0].get());
+                if (!hash) {
+                    return ast::MK::EmptyTree();
+                }
+                int i = -1;
+                for (auto &key : hash->keys) {
+                    i++;
+                    auto &value = hash->values[i];
+                    auto lit = ast::cast_tree<ast::Literal>(key.get());
+                    if (lit && lit->isSymbol(ctx)) {
+                        auto symName = lit->asSymbol(ctx);
+                        if (name == symName) {
+                            retType = value->deepCopy();
+                        }
+                    }
+                }
+                break;
+            }
+
+            case core::Names::abstract()._id: {
+                // Don't make this method at all since abstract methods can't
+                // have bodies
+                return nullptr;
+            }
+
+            case core::Names::override_()._id: {
+                // A total hack but we allow .void.void or .void.returns and
+                // the one with content wins
+                send->fun = core::Names::void_();
+            }
+        }
+        auto recv = ast::cast_tree<ast::Send>(send->recv.get());
+        send = recv;
+    }
+
+    send = ast::cast_tree<ast::Send>(sig->block->body.get());
+    while (send != nullptr) {
+        switch (send->fun._id) {
+            case core::Names::returns()._id: {
+                if (!retType) {
+                    return ast::MK::EmptyTree();
+                }
+                send->args[0] = move(retType);
+                break;
+            }
+
+            case core::Names::void_()._id: {
+                if (!retType) {
+                    return ast::MK::EmptyTree();
+                }
+                send->fun = core::Names::returns();
+                send->args.emplace_back(move(retType));
+                break;
+            }
+        }
+
+        auto recv = ast::cast_tree<ast::Send>(send->recv.get());
+        send = recv;
+    }
+    return expr;
+}
+
+void DefaultArgs::patchDSL(core::MutableContext ctx, ast::ClassDef *klass) {
+    vector<unique_ptr<ast::Expression>> newMethods;
+    ast::Send *lastSig = nullptr;
+    bool isOverload = false;
+
+    for (auto &stat : klass->rhs) {
+        typecase(
+            stat.get(),
+            [&](ast::Send *send) {
+                if (send->fun != core::Names::sig()) {
+                    return;
+                }
+                if (lastSig != nullptr) {
+                    isOverload = true;
+                    return;
+                }
+                lastSig = send;
+            },
+            [&](ast::MethodDef *mdef) {
+                if (isOverload) {
```

Comment:
Could you leave a comment in here about why we're skipping methods that have overloads? Reading through this code 6 months from now, I can imagine having to work back through why that's happening.